### PR TITLE
Add support for sparse complex tensors for CPU/CUDA

### LINKED ISSUE #50690: Add support for sparse complex tensors
## 🚀 Feature

Currently, sparse tensors only support real floating point tensors. Complex support should be added for completion.

## Motivation

I'm trying to use complex tensors for compressed sensing and am running out of GPU memory (4GiB); even though my tensors are sparse (~1%) I cannot use optimized routines, since PyTorch does not currently support them. I have resorted to converting them to scipy sparse matrices and going without autograd.

## Pitch

`torch.sparse.CFloatTensor`, `torch.sparse.CDoubleTensor` added as complex extensions of existing real data types.

## Status

Complex support was added in this PR #57125  for CPU/CUDA. 

- [x]  complex support (torch.cfloat and torch.cdouble) to torch.sparse_coo_tensor constructors
- [x]  complex support to coalesce function
- [x]  complex support to to_dense function
- [x]  complex support to to_sparse function
- [x]  complex support to sparse_add function
- [x]  complex support to addmm function
- [x]  unit tests

Note: That PR contains only complex support for torch.sparse_coo_tensor *forward/backward* function and the related ops used with this function (coalesce, to_dense, to_sparse, and sparse_add, addmm). The following PRs should cover other sparse operations to have a more complex sparse support, specifically related with the use of specific APIs for accelerated linear algebra. 
    
CSR sparse tensor: ops not implemented yet: 
- [ ] transpose, complex conjugate, torch.pow, torch.exp, torch.block_diag, torch.diagflat,  torch.eye 
- [ ]  matmult ops: bmm, sparse@sparse
- [x] Add complex support for CSR Sparse Tensors

cc @ezyang @anjali411 @dylanbespalko @mruberry @aocsa @nikitaved @pearu

### PR DESCRIPTION
**Pull request by aocsa**
Fixes #50690

Currently, sparse tensors only support real floating point tensors. Complex support is added in this PR for CPU/CUDA.

- [x] add complex support (torch.cfloat and torch.cdouble) to torch.sparse_coo_tensor constructors
- [x] add complex support to `coalesce`  function   
- [x] add complex support to `to_dense`  function   
- [x] add complex support to `to_sparse`   function
- [x] add complex support to `sparse_add`   function
- [x] add unit tests 

Note: This PR contains only complex support for torch.sparse_coo_tensor (fordward/backward) function and the related `ops` used with this function (`coalesce`, `to_dense`, `to_sparse`, and `sparse_add`).   The following PRs should cover other sparse operations to have a more complex sparse support, specifically related with the use of specific APIs for accelerated linear algebra using cuSparse and MKL.

### CODE REVIEW

**Review by pearu — CHANGES_REQUESTED** [review]
Looks good. This PR adds very basic support for complex COO tensors.

I think that the minimal complex sparse support should pass most of the sparse tests. I suggest applying the following patch

<details>

```
diff --git a/test/test_sparse.py b/test/test_sparse.py
index 41ebd60b41..ee1bd58d9f 100644
--- a/test/test_sparse.py
+++ b/test/test_sparse.py
@@ -48,6 +48,8 @@ def cuda_only(inner):
 
 class TestSparse(TestCase):
 
+    value_dtype = torch.float64
+
     def setUp(self):
         # These parameters control the various ways we can run the test.
         # We will subclass and override this method to implement CUDA
@@ -56,7 +58,6 @@ class TestSparse(TestCase):
         self.is_uncoalesced = False
         self.device = 'cpu'
         self.exact_dtype = True
-        self.value_dtype = torch.float64
         self.index_tensor = lambda *args: torch.tensor(*args, dtype=torch.int64, device=self.device)
         self.value_empty = lambda *args: torch.empty(*args, dtype=self.value_dtype, device=self.device)
         self.value_tensor = lambda *args: torch.tensor(*args, dtype=self.value_dtype, device=self.device)
@@ -3185,6 +3186,10 @@ class TestSparse(TestCase):
         self.assertRaises(TypeError, assign_to)
 
 
+class TestSparseComplex(TestSparse):
+    value_dtype = torch.complex128
+
+
 class TestUncoalescedSparse(TestSparse):
     def setUp(self):
         super(TestUncoalescedSparse, self).setUp()

```
</details>

to see what features are missing.

**Comment by aocsa** [comment]
> Looks good. This PR adds very basic support for complex COO tensors.
> 
> I think that the minimal complex sparse support should pass most of the sparse tests. I suggest applying the following patch
> 
> to see what features are missing.

I addressed the internal review from @pearu.  Specifically, this PR  fix  some errors in the tests when the new class  `TestSparseComplex` is enabled. I think this PR is ready to be merged and the following PRs should cover other sparse operations to have a minimal complex sparse support and should be done in other PRs because It will involve a lot of more changes, specifically related with the use of specific APIs for accelerated linear algebra using cuSparse and MKL.  cc 
@anjali411, @mruberry, @rgommers.    

Note: This is the last report when `TestSparseComplex` is fully enabled. 

```
======================================================================
test_dsmm (__main__.TestSparseComplex)
----------------------------------------------------------------------
    res = torch.dsmm(x, y)
======================================================================
test_hsmm (__main__.TestSparseComplex)
----------------------------------------------------------------------
    res = torch.hsmm(x, y)
======================================================================
test_pickle (__main__.TestSparseComplex)
    values = torch.arange(values_numel, dtype=self.value_dtype,

======================================================================
test_print (__main__.TestSparseComplex)
----------------------------------------------------------------------
    values = torch.arange(values_numel, dtype=self.value_dtype,
```

**Comment by anjali411** [comment]
@aocsa thanks for the PR! I'll try to review this PR this week but in the meantime please undo some third-party commits from your PR that you might have added by mistake :)

File: test/test_sparse.py [hunk 1]
```diff
@@ -3176,6 +3187,22 @@ def assign_to(a):
...
+
+    def test_hsmm(self):
+        self.skipTest("torch.hsmm skipped! Not implemented yet!")
+
+    def test_print(self):
+        self.skipTest("test_print skipped! Not implemented yet!")
+
+    def test_pickle(self):
+        self.skipTest("test_pickle skipped! Not implemented yet!")
+
```
Comment (pearu) [inline test/test_sparse.py:3205]: Thanks, @aocsa for these tests. I think this shows that sparse complex support is close to being complete, however, the `print` and `pickle` are pretty basic features. Can these be supported within this PR?

File: test/test_sparse.py [hunk 2]
```diff
@@ -3176,6 +3187,22 @@ def assign_to(a):
...
 class TestUncoalescedSparse(TestSparse):
```
Comment (pearu) [inline test/test_sparse.py:3206]: Could you also try:
```
class TestUncoalescedSparseComplex(TestSparseComplex):
    def setUp(self):
￼        super(TestUncoalescedSparseComplex, self).setUp()
        self.is_uncoalesced = True
```
?

Comment (pearu) [reply test/test_sparse.py:3206]: And similarly:
```
class TestCudaSparseComplex(TestSparseComplex):
    def setUp(self):
        ...
class TestCudaUncoalescedSparseComplex(TestCudaSparseComplex):
    def setUp(self):
        ...
```

File: test/test_sparse.py [hunk 3]
```diff
@@ -1275,13 +1277,18 @@ def _test_spadd_shape(self, nnz, shape_i, shape_v=None):
         nnz = i.size(1)
 
         # Non contiguous sparse indices tensor
-        x_ = self.sparse_tensor(i[:, ::2], v[:int(nnz / 2)], x.shape)
```
Comment (pearu) [inline test/test_sparse.py:1278]: A shorted fix would be:
```
x_ = self.sparse_tensor(i[:, ::2], v[:(nnz+1) // 2], x.shape)
```

File: test/test_sparse.py [hunk 4]
```diff
@@ -1275,13 +1277,18 @@ def _test_spadd_shape(self, nnz, shape_i, shape_v=None):
...
+        i_ = i[:, ::2]
+        new_nnz = i_.size(1)
+        v_ = v[:new_nnz]
+        x_ = self.sparse_tensor(i_, v_, x.shape)
         res = torch.add(y, x_, alpha=r)
         expected = y + r * self.safeToDense(x_)
         self.assertEqual(res, expected)
 
         # Non contiguous sparse values tensor
-        x_ = self.sparse_tensor(i[:, :int(nnz / 2)], v[::2], x.shape)
```
Comment (pearu) [inline test/test_sparse.py:1284]: Use:
```
x_ = self.sparse_tensor(i[:, :(nnz+1) // 2], v[::2], x.shape)
```

**Review by pearu — CHANGES_REQUESTED** [review]
Thanks, @aocsa  Looks good**2!

I'd like to propose implementing support for `print` and `pickle` for sparse complex tensors. In case this will require a (more complicated) fix for general complex tensors support, then nevermind as it could be done in a separate PR.

Also, I'd suggest testing sparse complex tensor support for uncoaleasced as well as CUDA device cases.

File: test/test_sparse.py (hunk 1 above)
Comment (aocsa) [reply test/test_sparse.py:3201]: `print` and `pickle` rely on `arange` op. This op is listed here. https://github.com/pytorch/pytorch/issues/33152 as a TODO and I think should be covered in their own PR.  

Moreover, I was looking the [arange_cpu] implementation(https://github.com/pytorch/pytorch/blob/270111b7b611d174967ed204776985cefca9c144/aten/src/ATen/native/RangeFactories.cpp#L161)  and it looks there are some issues related : 

https://github.com/pytorch/pytorch/issues/34266#issuecomment-725514202 (issue about remainder, floor, ceil on complex data)
https://github.com/pytorch/pytorch/issues/38775 (Linspace broken for complex dtypes )
https://github.com/numpy/numpy/issues/10332 (range behaves poorly on complex numbers)

Comment (pearu) [reply test/test_sparse.py:3201]: Yes, @aocsa I agree that the lack of complex support for these ops should not be addressed here.

**Review by pearu — APPROVED** [review]
LGTM. Thanks, @aocsa !

File: aten/src/ATen/native/sparse/cuda/SparseCUDATensorMath.cu [hunk 5]
```diff
@@ -436,10 +436,9 @@ SparseTensor& add_out_sparse_cuda(const SparseTensor& t, const SparseTensor& src
 
   Tensor t_values_ = t._values().to(commonDtype);
   Tensor s_values_ = src._values().to(commonDtype);
-
-  AT_DISPATCH_ALL_TYPES_AND2(
+  AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND2(
     at::ScalarType::Half, at::ScalarType::BFloat16, commonDtype, "add_out_sparse_cuda", [&] {
-      if (value.to<scalar_t>() != static_cast<scalar_t>(1)) {
+      if (value.to<scalar_t>() != scalar_t(1)) {
```
Comment (anjali411) [inline aten/src/ATen/native/sparse/cuda/SparseCUDATensorMath.cu:441]: We should continue using the `static_cast` here so please revert this change back

File: aten/src/TH/THBlas.cpp [hunk 6]
```diff
@@ -8,3 +8,6 @@
 
 #include <TH/generic/THBlas.cpp>
 #include <TH/THGenerateHalfType.h>
+
+#include <TH/generic/THBlas.cpp>
+#include <TH/THGenerateComplexTypes.h>
```
Comment (anjali411) [inline aten/src/TH/THBlas.cpp:13]: We have refrained from adding the TH code for complex so far barring the storage code which was blocking the autograd work. I think it's ok to add this however it's strongly preferable to migrate the functions from TH to ATen and then add complex support.

File: test/test_sparse.py [hunk 7]
```diff
@@ -2176,7 +2179,7 @@ def test_factory(self):
                 for use_tensor_idx in [True, False]:
                     for use_tensor_val in [True, False]:
                         for use_cuda in ([False] if not torch.cuda.is_available() else [True, False]):
-                            for dtype in [torch.float64, torch.float16]:
+                            for dtype in [torch.float64, torch.float16, torch.cfloat, torch.cdouble]:
```
Comment (anjali411) [inline test/test_sparse.py:2182]: it's odd to not just have `float` tested here so I guess we can add that too

File: test/test_sparse.py [hunk 8]
```diff
@@ -2323,6 +2326,10 @@ def test_factory_type_inference(self):
         self.assertEqual(torch.float32, t.dtype)
         t = torch.sparse_coo_tensor(torch.tensor(([0], [2])), torch.tensor([1.], dtype=torch.float64))
         self.assertEqual(torch.float64, t.dtype)
+        t = torch.sparse_coo_tensor(torch.tensor(([0], [2])), torch.tensor([1.], dtype=torch.cfloat))
```
Comment (anjali411) [inline test/test_sparse.py:2329]: consider having non zero imag value in the tensor e.g., `torch.tensor([2.3-4.5j], dtype=torch.cfloat)`

**Review by anjali411 — COMMENTED** [review]
General comments:
1. Addition of TH code logic for complex. How many other TH functions do you plan to enable for complex?
2. Needs tests for `coalesce`, `nonzero`, `add` and `to_dense` (forward and autograd).

File: aten/src/TH/THBlasUtils.h [hunk 9]
```diff
@@ -30,3 +30,4 @@ inline void THBlas_copy(int64_t n, T *x, int64_t incx, T *y, int64_t incy);
   }
 
 AT_FORALL_SCALAR_TYPES(COPY_SPECIALIZATION)
+AT_FORALL_COMPLEX_TYPES(COPY_SPECIALIZATION)
```
Comment (mruberry) [inline aten/src/TH/THBlasUtils.h:33]: Newline needed here.

File: aten/src/TH/generic/THBlas.cpp [hunk 10]
```diff
@@ -57,17 +68,23 @@ void THBlas_(copy)(int64_t n, scalar_t *x, int64_t incx, scalar_t *y, int64_t in
...
 
+
 #if defined(TH_REAL_IS_DOUBLE)
     dcopy_(&i_n, x, &i_incx, y, &i_incy);
-#else
+#elif defined(TH_REAL_IS_FLOAT)
     scopy_(&i_n, x, &i_incx, y, &i_incy);
+#elif defined(TH_REAL_IS_COMPLEXDOUBLE)
+    zcopy_(&i_n, x, &i_incx, y, &i_incy);
+#else
```
Comment (mruberry) [inline aten/src/TH/generic/THBlas.cpp:86]: Add a comment for what this case covers

File: aten/src/TH/generic/THBlas.cpp [hunk 11]
```diff
@@ -96,8 +114,12 @@ void THBlas_(axpy)(int64_t n, scalar_t a, scalar_t *x, int64_t incx, scalar_t *y
 
 #if defined(TH_REAL_IS_DOUBLE)
     daxpy_(&i_n, &a, x, &i_incx, y, &i_incy);
-#else
+#elif defined(TH_REAL_IS_FLOAT)
     saxpy_(&i_n, &a, x, &i_incx, y, &i_incy);
+#elif defined(TH_REAL_IS_COMPLEXDOUBLE)
+    zaxpy_(&i_n, &a, x, &i_incx, y, &i_incy);
+#else
```
Comment (mruberry) [inline aten/src/TH/generic/THBlas.cpp:121]: Add a comment for what this case covers

File: test/test_sparse.py [hunk 12]
```diff
@@ -298,7 +299,8 @@ def fn(x):
         ])
         # we don't have to_dense for half types on CPU because it is implemented
         # with a slower add_ operation
-        for dtype in [torch.float16, torch.float32, torch.float64] if self.device != 'cpu' else [torch.float32, torch.float64]:
+        for dtype in [torch.float16, torch.float32, torch.float64, torch.cfloat, torch.cdouble] \
```
Comment (mruberry) [inline test/test_sparse.py:302]: We really should invest in updating the sparse tests to use the device generic test framework so we can use the @dtypes decorator.

File: aten/src/ATen/LegacyTHFunctionsCPU.cpp [hunk 13]
```diff
@@ -99,6 +99,18 @@ Tensor & _th_nonzero_out(Tensor & result, const Tensor & self) {
             THBFloat16Tensor_nonzero(result_, self_);
             break;
         }
+        case ScalarType::ComplexDouble: {
```
Comment (anjali411) [inline aten/src/ATen/LegacyTHFunctionsCPU.cpp:102]: add tests for nonzero

File: aten/src/ATen/native/sparse/SparseTensorMath.cpp [hunk 14]
```diff
@@ -479,7 +479,7 @@ SparseTensor& add_out_sparse_contiguous(SparseTensor& r, const SparseTensor& t,
     auto r_indices_accessor = r_indices.accessor<int64_t, 2>();
     auto src_indices_accessor = src_indices.accessor<int64_t, 2>();
 
-    AT_DISPATCH_ALL_TYPES(
+    AT_DISPATCH_ALL_TYPES_AND_COMPLEX(
```
Comment (anjali411) [inline aten/src/ATen/native/sparse/SparseTensorMath.cpp:482]: add test

Comment (anjali411) [inline aten/src/ATen/native/sparse/SparseTensorMath.cpp:482]: please add tests!

File: aten/src/ATen/native/sparse/cuda/SparseCUDATensor.cu [hunk 15]
```diff
@@ -94,7 +94,7 @@ SparseTensor coalesce_sparse_cuda(const SparseTensor& self) {
     int64_t stride = at::prod_intlist(values.sizes().slice(1));
     dim3 grid(THCCeilDiv(newNnz, (int64_t) SZ), THCCeilDiv(stride, (int64_t) C10_WARP_SIZE*SZ));
     dim3 block(C10_WARP_SIZE, SZ);
-    AT_DISPATCH_ALL_TYPES_AND2(
+    AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND2(
```
Comment (anjali411) [inline aten/src/ATen/native/sparse/cuda/SparseCUDATensor.cu:97]: please add tests!

File: aten/src/ATen/native/sparse/cuda/SparseCUDATensorMath.cu (hunk 5 above)
Comment (aocsa) [reply aten/src/ATen/native/sparse/cuda/SparseCUDATensorMath.cu:441]: I tried to maintain that,  but when `static_cast` is enabled it doesn't compile with complex data type.

Comment (pearu) [reply aten/src/ATen/native/sparse/cuda/SparseCUDATensorMath.cu:441]: Just an idea, if `c10::complex` (I presume this corresponds to `scalar_t` if complex) would define a constructor using `int` as input, wouldn't `static_cast` work then?

Comment (pearu) [reply aten/src/ATen/native/sparse/cuda/SparseCUDATensorMath.cu:441]: See, for instance, https://stackoverflow.com/questions/7312745/static-cast-for-user-defined-types

File: aten/src/ATen/LegacyTHFunctionsCPU.cpp (hunk 13 above)
Comment (aocsa) [reply aten/src/ATen/LegacyTHFunctionsCPU.cpp:102]: done

File: aten/src/ATen/native/sparse/cuda/SparseCUDATensor.cu (hunk 15 above)
Comment (aocsa) [reply aten/src/ATen/native/sparse/cuda/SparseCUDATensor.cu:97]: All test for coalesce and uncoalesce cases with CPU and CUDA are covered in these new classes. 
`TestSparseComplex, ``TestUncoalescedSparseComplex`, `TestCudaSparseComplex`, `TestCudaUncoalescedSparseComplex`, and `TestUncoalescedSparse`
https://github.com/pytorch/pytorch/blob/5550442f7392a11c8f555109f33e884f30353b9b/test/test_sparse.py#L3203

**Comment by aocsa** [comment]
> General comments:

Thanks @anjali411 for the feedback. I addressed the reviews in this update.
 
> 1. Addition of TH code logic for complex. How many other TH functions do you plan to enable for complex?

I would like to maintain this PR short, so I don't have plans to add more logic. This PR contains only complex support for torch.sparse_coo_tensor (fordward/backward) function and the related `ops` used with this function (`coalesce`, `to_dense`, `to_sparse`, and `sparse_add`).   The following PRs should cover other sparse operations to have a more complex sparse support, specifically related with the use of specific APIs for accelerated linear algebra using cuSparse and MKL. 

> 2. Needs tests for `coalesce`, `nonzero`, `add` and `to_dense` (forward and autograd).

The following tests were added:  

New forward complex support tests to `test_sparse.py`:
- test_is_nonzero (nonzero)
- test_sparse_add_coalesce (add)
Note: `to_dense` and `coalesce were previously added 

New backward complex support  tests to `test_autograd.py`: 
- test_sparse_backward
- test_sparse_ctor_getter_backward

Looking forward to your comments. cc @anjali411, @mruberry

File: aten/src/ATen/native/sparse/cuda/SparseCUDATensorMath.cu (hunk 5 above)
Comment (mruberry) [reply aten/src/ATen/native/sparse/cuda/SparseCUDATensorMath.cu:441]: Looks like this still needs to be addressed?

File: test/test_autograd.py [hunk 16]
```diff
@@ -764,35 +764,35 @@ def forward(ctx, x, grad_x):
...
-        x = torch.randn(size, requires_grad=True)
-        (fn.apply(x, dense_grad) + fn.apply(x, sparse_grad1) + fn.apply(x, sparse_grad2)).sum().backward()
-        self.assertEqual(x.grad, dense_grad + sparse_grad1 + sparse_grad2)
-        # sparse only
-        x = torch.randn(size, requires_grad=True)
-        (fn.apply(x, sparse_grad1) + fn.apply(x, sparse_grad2)).sum().backward()
-        self.assertEqual(x.grad, sparse_grad1 + sparse_grad2)
+        for dtype in torch.testing.floating_and_complex_types():
+            size = torch.Size([6, 3, 2])
+            i1 = torch.LongTensor([
```
Comment (mruberry) [inline test/test_autograd.py:769]: I know this is in older code, but while we're here let's not use the deprecated torch.LongTensor and use torch.tensor constructors (here and on line 775 below)

File: test/test_autograd.py [hunk 17]
```diff
@@ -764,35 +764,35 @@ def forward(ctx, x, grad_x):
...
-        self.assertEqual(x.grad, dense_grad + sparse_grad1 + sparse_grad2)
-        # dense first
-        x = torch.randn(size, requires_grad=True)
-        (fn.apply(x, dense_grad) + fn.apply(x, sparse_grad1) + fn.apply(x, sparse_grad2)).sum().backward()
-        self.assertEqual(x.grad, dense_grad + sparse_grad1 + sparse_grad2)
-        # sparse only
-        x = torch.randn(size, requires_grad=True)
-        (fn.apply(x, sparse_grad1) + fn.apply(x, sparse_grad2)).sum().backward()
-        self.assertEqual(x.grad, sparse_grad1 + sparse_grad2)
+        for dtype in torch.testing.floating_and_complex_types():
```
Comment (mruberry) [inline test/test_autograd.py:767]: Moving this test to TestAutogradDeviceType will let you make it a generic device framework test template, then it can use the @dtypes decorator and be instantiated for both CUDA and CPU.

File: test/test_autograd.py [hunk 18]
```diff
@@ -6668,10 +6669,10 @@ def _test(size, sparse_dim, nnz, device):
 
             inp = torch.randn(v_size, requires_grad=True)
             other = self.genSparseTensor(size, sparse_dim, nnz, is_uncoalesced=True)[0]
-            other = other.to(device)
+            other = other.to(device).to(dtype)
```
Comment (mruberry) [inline test/test_autograd.py:6672]: `.to(device=device, dtype=dtype)`

File: test/test_autograd.py [hunk 19]
```diff
@@ -6658,7 +6658,8 @@ def test_parameter_resize(self, device):
             m = torch.cat((asd, asd))
             m.sum().backward()
 
-    def test_sparse_ctor_getter_backward(self, device):
+    @dtypes(torch.double, torch.cdouble)
+    def test_sparse_ctor_getter_backward(self, device, dtype):
```
Comment (mruberry) [inline test/test_autograd.py:6662]: This test class is where I suggest moving the other test to.

File: test/test_sparse.py [hunk 20]
```diff
@@ -49,6 +49,8 @@ def outer(self, *args, **kwargs):
 
 class TestSparse(TestCase):
 
+    value_dtype = torch.float64
```
Comment (mruberry) [inline test/test_sparse.py:52]: Why this change?

File: test/test_sparse.py [hunk 21]
```diff
@@ -3177,12 +3195,53 @@ def assign_to(a):
...
+class TestCudaSparseComplex(TestSparseComplex):
+
+    def setUp(self):
+        super(TestCudaSparseComplex, self).setUp()
+        self.is_cuda = True
+        self.device = 'cuda'
+
+
+@unittest.skipIf(not TEST_CUDA, 'CUDA not available')
+class TestCudaUncoalescedSparseComplex(TestCudaSparseComplex):
```
Comment (mruberry) [inline test/test_sparse.py:3238]: We should really modernize test_sparse.py to use the device generic test framework (and OpInfos) and simplify this variant instantiation process.

**Review by mruberry — COMMENTED** [review]
Hey @mattip! I took another look. Overall this is looking good. I think tests for complex nonzero still need to be added, though?

The current nonzero tests are here:

https://github.com/pytorch/pytorch/blob/de7eeb7752d4bd599901a9adcafbfc4a3ea25035/test/test_shape_ops.py#L507

File: test/test_sparse.py (hunk 20 above)
Comment (pearu) [reply test/test_sparse.py:52]: @mruberry this was introduced for reusing `TestSparse` tests for complex dtype. See https://github.com/pytorch/pytorch/pull/50984/files#diff-bf009aed03d333c6a63253630fed4465ba5f7e025029fe2a22919759c38beb6dR3198-R3201

**Comment by mattip** [comment]
> Hey @mattip! I took another look...

@mruberry I think you meant @aocsa

**Comment by mruberry** [comment]
> > Hey @mattip! I took another look...
> 
> @mruberry I think you meant @aocsa

Sorry @mattip, I did mean @aocsa. Looking at too many PRs at the same time!

File: test/test_sparse.py (hunk 21 above)
Comment (aocsa) [reply test/test_sparse.py:3238]: I'm working on this, I think this task really worth. Using the new generic test framework I've found out some issues like misuses of the API and some memory leaks.

**Comment by aocsa** [comment]
Thanks @mruberry for the feedback.  I took a time to  modernize  `test_sparse.py` to use the new device generic test framework. I added  a new parameter to `instantiate_device_type_tests` to simplify  the variant instantiation process for coalesced/uncoalesced sparse tensors too.

Moreover, in the process I've found out some issues.  
1. Some test didn't  completely support some dtypes  as the instance variable `self.value_dtype` were not used uniformly in all the tests. e.g.  
  https://github.com/Quansight/pytorch/blob/d648f312d99a932de396ef409c57c8f8ae9ba344/test/test_sparse.py#L2078
Note: For example, with the use of `@dtype` decorator now is clear which test support complex data or not. 

2. A couple of tests have some cuda memory leaks for the backward algorithm. 
- `test_sparse_sum`, `test_sparse_matmul`. 
Note: I created a new issue here   https://github.com/pytorch/pytorch/issues/52253. I think this issue was is reported now because with the new variant instantiation process, all CUDA tests have CUDATestBase as base class and with this there are specific routines to check memory usage before and after running a test .  

> Hey @mattip! I took another look. Overall this is looking good. I think tests for complex nonzero still need to be added, though?
> 
> The current nonzero tests are here:
> 
> https://github.com/pytorch/pytorch/blob/de7eeb7752d4bd599901a9adcafbfc4a3ea25035/test/test_shape_ops.py#L507

Tests for complex support for `nonzero` is here:  https://github.com/Quansight/pytorch/blob/d648f312d99a932de396ef409c57c8f8ae9ba344/test/test_sparse.py#L2632

Looking forward your comments cc @ngimel, @mruberry

File: aten/src/TH/THBlas.cpp (hunk 6 above)
Comment (ngimel) [reply aten/src/TH/THBlas.cpp:13]: @aocsa please port swap, copy and axpy functions to ATen before adding complex support to them (can be done in a separate PR, and this can be rebased on top). There is absolutely no reason to add complex support to TH in this case.

Comment (aocsa) [reply aten/src/TH/THBlas.cpp:13]: I create a new PR to port `copy` and `axpy` functions, I didn't port `swap` as it is not used anywhere.  
https://github.com/pytorch/pytorch/pull/52345

File: aten/src/ATen/native/sparse/cuda/SparseCUDATensorMath.cu (hunk 5 above)
Comment (aocsa) [reply aten/src/ATen/native/sparse/cuda/SparseCUDATensorMath.cu:441]: I made the changes where it was possible, I was able to use `static_cast` with complex dtype when `cpp` compiler is used, however when `nvcc` compiler is used  this doesn't work, I tried to use the strategy suggested by pearu but it generate conflicts with other functions. So I guess the issue is with nvcc compiler as in other cases  😔

**Comment by aocsa** [comment]
This PR is rebased. Now it has the changes of PR #52345 (blas function to aten).  Moreover, about the CUDA memory leaks, PR #52874 is solving this issue, I can wait to be merged to rebase this PR. Besides that, 
let me know if anything needs to be taken care of from my end to merge this into PR. cc @mruberry, @ngimel

**Comment by aocsa** [comment]
> This PR is rebased. Now it has the changes of PR #52345 (blas function to aten). Moreover, about the CUDA memory leaks, PR #52874 is solving this issue. Besides that, let me know if anything needs to be taken care of from my end to merge this into PR. cc @mruberry, @ngimel

This PR  is too big so ghstack is  used now and this PR will be closed.
The following PRs at #54153  will add complex support @mruberry, @ngimel, @anjali411
